### PR TITLE
Remove IronPython.Compiler.Ast.Arg

### PR DESCRIPTION
**Pull request by slozier**
Gets rid of the annoying (to me) `Arg` construct. :smile:

@BCSharp if you have any feedback. Not sure if this will creates merge conflicts with your PR.

### CODE REVIEW

**Comment by slozier** [comment]
Looks like there are some issues with the start/end indices during parsing.

**Comment by slozier** [comment]
Looks like I also broke:
```Python
class test(*[object]): pass
```

**Comment by BCSharp** [comment]
> @BCSharp if you have any feedback.

I don't care about `Arg` that much. There are other things that bug me way more.

>  Not sure if this will creates merge conflicts with your PR.

I does create a merge conflict in `CallExpression` that is easilly fixable. More importantly, however, the code will not compile because I still use `IReadOnlyList<Arg> Keywords` in `ClassDefinition`.

File: Src/IronPython/Compiler/Ast/SequenceExpression.cs [hunk 1]
```diff
@@ -18,13 +18,15 @@
...
+        protected readonly Expression[] _items;
 
         protected SequenceExpression(Expression[] items) {
             _items = items;
         }
 
-        public IList<Expression> Items => _items;
+        public IReadOnlyList<Expression> Items => _items;
+
+        internal Expression[] UnsafeItems => _items; // TODO: get rid of this
```
Comment (BCSharp) [inline Src/IronPython/Compiler/Ast/SequenceExpression.cs:29]: Good luck with this! 😄  `IReadOnlyCollection` should have had `CopyTo`, just like `ICollection` has. I suppose `ArrayUtils.Insert<T>(T item, IReadOnlyList<T> list)` can also do `if (list is T[] array) ...` to maintain performance .

**Comment by slozier** [comment]
@BCSharp Thanks for the review! All excellent points. I guess I was a bit quick on the merge. 😄

File: Src/IronPython/Compiler/Ast/CallExpression.cs [hunk 2]
```diff
@@ -107,69 +114,68 @@ public class CallExpression : Expression, IInstructionProvider {
...
 
                 var expressions = new ReadOnlyCollectionBuilder<MSAst.Expression>(numDict + 2);
                 var varExpr = Expression.Variable(typeof(PythonDictionary), "$dict");
                 expressions.Add(Expression.Assign(varExpr, Expression.Call(AstMethods.MakeEmptyDict)));
-                for (int i = firstDictPos; i < kwargs.Count; i++) {
-                    var arg = kwargs[i];
-                    if (arg.ArgumentInfo.Kind == ArgumentType.Dictionary) {
+                foreach (var arg in kwargs) {
+                    if (arg.Name is null) {
                         expressions.Add(Expression.Call(AstMethods.DictMerge, context, varExpr, arg.Expression));
```
[the remark this answers is not in the document]
Comment (slozier) [reply Src/IronPython/Compiler/Ast/CallExpression.cs:179]: I had noticed the difference and tried a few thing to make it fail, but without success so I left it as is. I guess I wasn't under scope isolation. Will add it, better safe than sorry I guess.

File: Src/IronPython/Compiler/Ast/ClassDefinition.cs [hunk 3]
```diff
@@ -37,10 +37,10 @@ public class ClassDefinition : ScopeStatement {
         private static readonly MSAst.ParameterExpression _parentContextParam = Ast.Parameter(typeof(CodeContext), "$parentContext");
         private static readonly MSAst.Expression _tupleExpression = MSAst.Expression.Call(AstMethods.GetClosureTupleFromContext, _parentContextParam);
 
-        public ClassDefinition(string name, IReadOnlyList<Arg>? bases, IReadOnlyList<Arg>? keywords, Statement? body = null) {
+        public ClassDefinition(string name, IReadOnlyList<Expression>? bases, IReadOnlyList<Keyword>? keywords, Statement? body = null) {
             _name = name;
-            _bases = bases?.ToArray() ?? Array.Empty<Arg>();
-            _keywords = keywords?.ToArray() ?? Array.Empty<Arg>();
+            _bases = bases?.ToArray() ?? Array.Empty<Expression>();
```
[the remark this answers is not in the document]
Comment (slozier) [reply Src/IronPython/Compiler/Ast/ClassDefinition.cs:42]: I don't mind changing this, but I don't know if there's any performance benefit to it? Presumably the jit makes the call disappear, but I don't know enough about this to be sure.

Looks like there are currently 0 uses of `Expression.EmptyArray` so if we decide to keep using `Array.Empty<Expression>()` I would remove that instead.

File: Src/IronPython/Compiler/Ast/SequenceExpression.cs (hunk 1 above)
Comment (slozier) [reply Src/IronPython/Compiler/Ast/SequenceExpression.cs:29]: Unfortunately `ArrayUtils.Insert` is in the DLR and I didn't want to rev it just for this. I guess as long as we don't start using `UnsafeItems` all over the place (I don't see why we would) this is fine.

File: Src/IronPython/Compiler/Ast/ClassDefinition.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (BCSharp) [reply Src/IronPython/Compiler/Ast/ClassDefinition.cs:42]: I would go with removing `Expression.EmptyArray`. Have simply noticed it was there so supposedly was meant to be used, but actually I prefer `Array.Empty<>`.

**Pull request merged**